contracts/FeralfileArtworkV4_2.sol

```diff
@@ -0,0 +1,184 @@
+
+    mapping(uint256 => uint256) private seriesNextSaleTokenIds; // seriesID -> tokenID
+
+    constructor(
+        string memory name_,
+        string memory symbol_,
+        bool burnable_,
+        bool bridgeable_,
+        address signer_,
+        address vault_,
+        address costReceiver_,
+        string memory contractURI_,
+        uint256[] memory seriesIds_,
+        uint256[] memory seriesMaxSupplies_,
+        uint256[] memory seriesNextSaleTokenIds_
+    )
+        FeralfileExhibitionV4_1(
+            name_,
+            symbol_,
+            burnable_,
+            bridgeable_,
+            signer_,
+            vault_,
+            costReceiver_,
+            contractURI_,
+            seriesIds_,
+            seriesMaxSupplies_
+        )
+    {
+        VAULT_V2 = IFeralfileVaultV2(payable(vault_));
+        for (uint256 i = 0; i < seriesIds_.length; i++) {
+            seriesNextSaleTokenIds[seriesIds_[i]] = seriesNextSaleTokenIds_[i];
+        }
+    }
+
+    /// @notice pay to get artworks to a destination address. The pricing, costs and other details is included in the saleData
+    /// @param r_ - part of signature for validating parameters integrity
+    /// @param s_ - part of signature for validating parameters integrity
+    /// @param v_ - part of signature for validating parameters integrity
+    /// @param saleData_ - the sale data
+    function buyBulkArtworks(
+        bytes32 r_,
+        bytes32 s_,
+        uint8 v_,
+        SaleDataV2 calldata saleData_
+    ) external payable {
+        require(_selling, "FeralfileExhibitionV4: sale is not started");
+        super._checkContractOwnedToken();
+        validateSaleDataV2(saleData_);
+
+        saleData_.payByVaultContract
+            ? VAULT_V2.payForSaleV2(r_, s_, v_, saleData_)
+            : require(
+                saleData_.price == msg.value,
+                "FeralfileExhibitionV4: invalid payment amount"
+            );
+
+        bytes32 message = keccak256(
+            abi.encode(block.chainid, address(this), saleData_)
+        );
+
+        //check nonce
+        _useCheckedNonce(saleData_.destination, saleData_.nonce);
+
+        if (!isValidSignature(message, r_, s_, v_)) {
+            revert InvalidSignature();
+        }
+
+        uint256 itemRevenue;
+        if (saleData_.price > saleData_.cost) {
+            itemRevenue =
+                (saleData_.price - saleData_.cost) /
+                saleData_.quantity;
+        }
+
+        uint256 nextSaleTokenId = seriesNextSaleTokenIds[saleData_.seriesID];
+        uint256 i = 0;
+        while (i < saleData_.quantity) {
+            uint256 tokenIdForSale = nextSaleTokenId;
+
+            if (!_exists(tokenIdForSale)) {
+                revert TokenIDNotFound();
+            }
+
+            nextSaleTokenId++;
+            if (ownerOf(tokenIdForSale) != address(this)) {
+                continue;
+            }
+
+            // send NFT
+            _safeTransfer(
```

Comment:
There would be a case that the remaining purchasable tokens is not enough compare to the requested quantity, the tx would be failed eventually but it will cost much gas since we transferred the token and emitted event already. So i'd suggest to accumulate the purchasable tokens into an array then loop over them to transfer.

Reply:
this is a rare case that only happens when out of tokens. but storing tokens in an array and performing an additional loop would cost some additional gas for each transaction. Since it's a rare case, so I think it would be more efficient to keep the transfer inside this loop.